src/hnsw/validate_index.c

```diff
@@ -0,0 +1,650 @@
+             level,
+             nodes_per_level[ level ],
+             edges_per_level[ level ],
+             min_neighbors_per_level[ level ],
+             max_neighbors_per_level[ level ]);
+    }
+    pfree(max_neighbors_per_level);
+    pfree(min_neighbors_per_level);
+    pfree(edges_per_level);
+    pfree(nodes_per_level);
+}
+
+void ldb_vi_free_neighbors(struct ldb_vi_node *vi_nodes, uint32 nodes_nr)
+{
+    for(uint32 i = 0; i < nodes_nr; ++i) {
+        struct ldb_vi_node *node = &vi_nodes[ i ];
+
+        for(uint32 level = 0; level <= node->vn_level; ++level) pfree(node->vn_neighbors[ level ]);
+        pfree(node->vn_neighbors);
+        pfree(node->vn_neighbors_nr);
+    }
+}
+
+void ldb_validate_index(Oid indrelid, bool print_info)
+{
+    Relation             index;
+    BlockNumber          header_blockno = 0;
+    Buffer               header_buf;
+    Page                 header_page;
+    HnswIndexHeaderPage *index_header;
+    MemoryContext        memCtx;
+    MemoryContext        saveCtx;
+    BlockNumber          blocks_nr;
+    uint32               nodes_nr;
+    struct ldb_vi_block *vi_blocks;
+    struct ldb_vi_node  *vi_nodes;
+
+    /* the code here doesn't change the index, so AccessShareLock is enough */
+    index = relation_open(indrelid, AccessShareLock);
+
+    if(print_info) {
+        elog(INFO, "validate_index() start for %s with Oid=%u", RelationGetRelationName(index), indrelid);
+    } else {
+        elog(INFO, "validate_index() start for %s", RelationGetRelationName(index));
+    }
+    memCtx = AllocSetContextCreate(CurrentMemoryContext, "hnsw validate_index context", ALLOCSET_DEFAULT_SIZES);
+    saveCtx = MemoryContextSwitchTo(memCtx);
+
+    header_buf = ReadBuffer(index, header_blockno);
+    LockBuffer(header_buf, BUFFER_LOCK_EXCLUSIVE);
+    header_page = BufferGetPage(header_buf);
+    index_header = (HnswIndexHeaderPage *)PageGetContents(header_page);
+    if(index_header->magicNumber != LDB_WAL_MAGIC_NUMBER) {
+        elog(ERROR,
+             "Invalid HnswIndexHeaderPage.magicNumber (page %" PRIu32 ", got %x, expected %x)",
+             header_blockno,
+             index_header->magicNumber,
+             LDB_WAL_MAGIC_NUMBER);
+    }
+    if(print_info) {
+        elog(INFO,
+             "index_header = HnswIndexHeaderPage("
+             "version=%" PRIu32 " vector_dim=%" PRIu32 " m=%" PRIu32 " ef_construction=%" PRIu32 " ef=%" PRIu32
+             " metric_kind=%d num_vectors=%" PRIu32 " last_data_block=%" PRIu32 " blockmap_page_groups=%" PRIu32 ")",
+             index_header->version,
+             index_header->vector_dim,
+             index_header->m,
+             index_header->ef_construction,
+             index_header->ef,
+             index_header->metric_kind,
+             index_header->num_vectors,
+             index_header->last_data_block,
+             index_header->blockmap_page_groups);
+    }
+
+    blocks_nr = RelationGetNumberOfBlocksInFork(index, MAIN_FORKNUM);
```

Comment:
> Could this be INIT_FORKNUM, once we implement support for unlogged tables?

It could. I can add the fork number as a parameter for this function. Or check for INIT fork and use it if it exists. What approach would you like to see here?

> is it possible to read from the opened relation whether the relation is on MAIN or INIT fork?

It is. It's just a matter of passing a parameter across the code. Do you want me to add the parameter?

Reply:
no. it is good as is for now! just confirming that adding unlogged tables would not add a lot of work for this function